src/storage/BlockStorage.ts

```diff
@@ -0,0 +1,284 @@
+            const block = Codec.decode(
+                blockConfirmation.signedBlock.encodedBlock,
+                Type.Block
+            );
+            const { forkId, height } = BlockUtils.getCoordinates(block);
+            const coordinateKey = this.coordinatesToKey([forkId, height]);
+
+            this.hashToBlockMap.delete(hashOrForkId as Hash);
+            this.coordinatesToBlockMap.delete(coordinateKey);
+            return true;
+        }
+
+        // ┌─ ROUTES TO: [OVERLOAD 2] - delete by coordinates
+        const coordinateKey = this.coordinatesToKey([
+            hashOrForkId as ForkId,
+            height
+        ]);
+        const blockConfirmation = this.coordinatesToBlockMap.get(coordinateKey);
+        if (!blockConfirmation) return false;
+
+        // Need to find and delete from hash map too
+        const blockHash = ethers.keccak256(
+            blockConfirmation.signedBlock.encodedBlock
+        );
+
+        this.coordinatesToBlockMap.delete(coordinateKey);
+        this.hashToBlockMap.delete(blockHash);
+        return true;
+    }
+
+    // ====================================
+    // PRIVATE HELPERS
+    // ====================================
+
+    private coordinatesToKey(coordinates: ForkCoordinates): CoordinateKey {
+        const [forkId, height] = coordinates;
+        return `${forkId}:${height}`;
+    }
+
+    private storeBlockConfirmation(
+        blockConfirmation: BlockConfirmationStruct
+    ): Hash {
+        const blockHash = ethers.keccak256(
+            blockConfirmation.signedBlock.encodedBlock
+        );
+        const block = Codec.decode(
+            blockConfirmation.signedBlock.encodedBlock,
+            Type.Block
+        );
+        const { forkId, height } = BlockUtils.getCoordinates(block);
+
+        return this.storeBlockConfirmationWithKeys(
+            blockConfirmation,
+            blockHash,
+            [forkId, height]
+        );
+    }
+
+    private storeBlockConfirmationWithKeys(
+        blockConfirmation: BlockConfirmationStruct,
+        blockHash: Hash,
+        coordinates: ForkCoordinates
+    ): Hash {
+        const coordinateKey = this.coordinatesToKey(coordinates);
+
+        // Check for duplicates - throw if exists
+        if (this.hashToBlockMap.has(blockHash)) {
+            throw new Error(`Block with hash ${blockHash} already exists`);
+        }
+        if (this.coordinatesToBlockMap.has(coordinateKey)) {
+            throw new Error(
+                `Block at fork ${coordinates[0]}, height ${coordinates[1]} already exists`
+            );
+        }
+
+        this.hashToBlockMap.set(blockHash, blockConfirmation);
+        this.coordinatesToBlockMap.set(coordinateKey, blockConfirmation);
+        return blockHash;
+    }
+
+    private isBlockConfirmation(
+        blockData: SignedBlockStruct | BlockConfirmationStruct
+    ): blockData is BlockConfirmationStruct {
+        return "signedBlock" in blockData && "signatures" in blockData;
```

Comment:
we should include a test for just this, so if `BlockConfirmation` ever changes - the test will fail - so we don't forget to update this if needed

Reply:
yes, overloading is over engineered and does not work well in ts

Reply:
broke it down to 2 ovelaoded functions
deleted `isBlockConfirmation` function